Update release checklist. (#10000)
ethereum/solidity: ReleaseChecklist.md

### Create the Release
 - [ ] Create Github release page: https://github.com/ethereum/solidity/releases/new
 - [ ] On the release page, select the ``develop`` branch as new target and set tag to the new version (e.g. `v0.5.4`) (make sure you only `SAVE DRAFT` instead of `PUBLISH RELEASE` before the actual release)

>  - [ ] Thank voluntary contributors in the Github release page (use ``git shortlog -s -n -e v0.5.4..origin/develop``).

Reviewer: Maybe we should use the placeholder here too to avoid someone blindly using 0.5.4?
```suggestion
 - [ ] Thank voluntary contributors in the Github release page (use ``git shortlog -s -n -e v$VERSION..origin/develop``).
```

Author: should be "previous version"

Reviewer: Ah, right. My mistake. So `$PREVIOUS_VERSION`?

Reviewer: But in that case  `v0.5.4` was a mistake too. Right above it says "set tag to the new version (e.g. `v0.5.4`)".

Author: Right! Will change to v0.6.3

 - [ ] Make a final check that there are no platform-dependency issues in the ``solidity-test-bytecode`` repository.
 - [ ] Check that all tests on the latest commit in ``develop`` are green.
 - [ ] Click the `PUBLISH RELEASE` button on the release page, creating the tag.

>  - [ ] Wait for the CI runs on the tag itself (travis will push the source archive and the static linux binary onto the Github release page).

Reviewer: So Travis pushes the source tarball now? Or was the checklist wrong? I don't see any change in Travis config and last time I asked you said the tarball is created manually...

Author: In the previous releases, I noticed that the tarball hash changed some time. Now I realized it was because it is pushed again by travis and thus just overwritten. We can change that once we drop travis altogether.

Reviewer: ok. good to know.

### Download Binaries
 - [ ] Take the ``solc.exe`` binary from the ``b_win_release`` run of the released commit in circle-ci and add it to the release page as ``solc-windows.exe``.
 - [ ] Take the ``solc`` binary from the ``b_osx`` run of the released commit in circle-ci and add it to the release page as ``solc-macos``.
 - [ ] If not done by travis: Take the ``soljson.js`` binary from the ``b_ems`` run of the released commit in circle-ci and add it to the release page as ``soljson.js``.

> ### Update solc-bin

Reviewer: Not really necessary as long as you're the one making releases but in case anyone else has to do it a link to solc-bin would make it easier to find out that it's a separate repo if one does not know that yet :)

```suggestion
### Update [solc-bin](https://github.com/ethereum/solc-bin/)
```

 - [ ] Copy ``soljson.js`` to ``solc-bin/bin/soljson-v$VERSION+commit.$COMMIT.js``
 - [ ] Copy ``solc-static-linux`` from the release page to ``solc-bin/linux-amd64/solc-linux-amd64-v$VERSION+commit.$COMMIT``
 - [ ] Make it executable.

>  - [ ] Copy ``solc-macos`` from the release page to ``solc-bin/macos-amd64/solc-macos-amd64-v$VERSION+commit.$COMMIT``

Reviewer: ```suggestion
 - [ ] Copy ``solc-macos`` from the release page to ``solc-bin/macosx-amd64/solc-macosx-amd64-v$VERSION+commit.$COMMIT``
```

 - [ ] Make it executable.

>  - [ ] Copy ``solc-windows.zip`` from the release page to ``solc-bin/windows-amd64/solc-windows-amd64-v$VERSION+commit.$COMMIT.zip``

Reviewer: ```suggestion
 - [ ] Copy ``solc-windows.exe`` from the release page to ``solc-bin/windows-amd64/solc-windows-amd64-v$VERSION+commit.$COMMIT.exe``
```

 - [ ] Run ``./update --reuse-hashes`` in ``solc-bin`` and verify that the script has updated ``list.js``, ``list.txt`` and ``list.json`` files correctly and that symlinks to the new release have been added in ``solc-bin/wasm/`` and ``solc-bin/emscripten-wasm32/``.
 - [ ] Create a pull request and merge.

### Homebrew and MacOS

>  - [ ] Update the version and the hash (``sha256sum solidity_x.x.x.tar.gz``) in https://github.com/Homebrew/homebrew-core/blob/master/Formula/solidity.rb
>  - [ ] Update the version and the hash (``sha256sum solidity_x.x.x.tar.gz``) in https://github.com/ethereum/homebrew-ethereum/blob/master/solidity.rb

Reviewer: ```suggestion
 - [ ] Update the version and the hash (``sha256sum solidity_$VERSION.tar.gz``) in https://github.com/Homebrew/homebrew-core/blob/master/Formula/solidity.rb
 - [ ] Update the version and the hash (``sha256sum solidity_$VERSION.tar.gz``) in https://github.com/ethereum/homebrew-ethereum/blob/master/solidity.rb
```

### Docker
 - [ ] Run ``./scripts/docker_deploy_manual.sh v$VERSION``).

### PPA
 - [ ] Change ``scripts/release_ppa.sh`` to match your key's email and key id.
 - [ ] Run ``scripts/release_ppa.sh v$VERSION`` to create the PPA release (you need the relevant openssl key).
 - [ ] Wait for the ``~ethereum/ubuntu/ethereum-static`` PPA build to be finished and published for *all platforms*. SERIOUSLY: DO NOT PROCEED EARLIER!!! *After* the static builds are *published*, copy the static package to the ``~ethereum/ubuntu/ethereum`` PPA for the destination series ``Trusty`` and ``Xenial`` while selecting ``Copy existing binaries``.

### Documentation

>  - [ ] Build the new version on https://readthedocs.org/projects/solidity/ (select `latest` on the bottom of the page and click `BUILD`)

Reviewer: ```suggestion
 - [ ] Build the new version on https://readthedocs.org/projects/solidity/ (select `latest` at the bottom of the page and click `BUILD`)
```

 - [ ] In the admin panel, select `Versions` in the menu and set the default version to the released one.

### Release solc-js

>  - [ ] Wait until solc-bin was properly deployed. You can test this via remix - a test run through remix is advisable anyway.

Reviewer: You mean until the files become available at `https://solc-bin.ethereum.org`?
Then maybe visiting `https://solc-bin.ethereum.org/bin/soljson-v$VERSION+commit.$COMMIT.js` as a sanity check is also a good idea.

Also, this might be a bit too specific to add to the list but making sure that the [S3 sync action](https://github.com/ethereum/solc-bin/actions?query=workflow%3A"Mirror+repository+content+to+an+S3+bucket") did not fail might be a good additional check. Or at least it will give some feedback on the progress of the sync.

Author: I think the test via remix is better. You can also do a simple compilation run and retrieve the version reported by the compiler.

Reviewer: Sure. Just a suggestion.

 - [ ] Increment the version number, create a pull request for that, merge it after tests succeeded.
 - [ ] Run ``npm publish`` in the updated ``solc-js`` repository.

>  - [ ] Create the tag ``git tag v$VERSION`` and push it with ``git push --tags``.

Reviewer: An annotated tag might be a good idea. It has a creation date and `git describe` by default skips non-annotated tags. I think annotated tags are recommended for tags representing versions. The downside is that you need to supply a message (like for a commit).

```suggestion
 - [ ] Create a tag using ``git tag --annotate v$VERSION`` and push it with ``git push --tags``.
```
